go/login/login.go

```diff
@@ -0,0 +1,306 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package login
+
+import (
+	"encoding/base64"
+	"fmt"
+	"github.com/google/glome/go/glome"
+	"net/url"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+const (
+	// Minimal acceptable length of a handshake. 1 byte for the prefix, 32 bytes for the key
+	MinHandshakeLen = 1 + glome.PublicKeySize
+)
+
+var (
+	// ErrFailedHandshake denotes that the URL has a wrong format.
+	ErrInvalidURLFormat = fmt.Errorf("url is malformed")
+	// ErrFailedHandshake denotes that the handshake is too short.
+	ErrInvalidHandshakeLen = fmt.Errorf("handshake length is small: should be at least %d", MinHandshakeLen)
+	// ErrVersionNotSupported denotes that the version of glome-login URL format is not supported.
+	ErrVersionNotSupported = fmt.Errorf("version not supported")
+	// ErrInvalidPrefixType denotes that the prefix type is invalid.
+	ErrInvalidPrefixType = fmt.Errorf("invalid prefix type")
+)
+
+// Contains the <message> information
```

Comment:
I think this comment should elaborate what a Message is and what it represents rather than this generic statement.

Reply:
Fixed.